src/Npgsql/Replication/PgOutput/PgOutputReplicationOptions.cs

```diff
@@ -43,12 +45,13 @@ public PgOutputReplicationOptions(IEnumerable<string> publicationNames, ulong pr
             ProtocolVersion = protocolVersion;
             Binary = binary;
             Streaming = streaming;
+            Messages = messages;
         }
 
         /// <summary>
         /// The version of the logical streaming replication protocol
         /// </summary>
-        public ulong ProtocolVersion { get; }
+        public ulong? ProtocolVersion { get; }
```

Comment:
If I follow things correctly, when this is null, we don't send a requested version to PG when starting the protocol. So right now, we default to protocol version 1 (to maintain backwards compat for existing applications which don't know about the new messages), but allow the user to either explicitly ask for version 2, or say null to mean "newest"? Is that right?

Reply:
Almost.
The problem is, that at the stage where the user sets up the PgOutputReplicationOptions we are not in the position to offer a useful default for the protocol version based on the backend he/she is going to connect to. We simply don't have a reference to an open connection there.
That's why we need to detect if a specific protocol version was explicitly requested and if that isn't the case, use the newest protocol version supported by both Npgsql and the backend when replication is started.

Reply:
OK. So we basically want Npgsql to always default to protocol v1 - regardless of what the user's PG server supports - in order to avoid breaking changes from new messages introduced in later versions? If so, then do we even think it makes sense to allow users to specify null? Doesn't that make their application inherently breakable when PG is upgraded?

In other words, it may be better to default to protocol version 1 (as now), and to *not* accept null as a possibility - this way the user always must specify a specific version which they support... What do you think?

Reply:
No, what I had in mind was to always default to the latest protocol supported by the backend unless it is specified otherwise in the options.

Reply:
The problem is that at least with v1 and v2, applications coded against PG13 would break immediately once executed against PG14 right? In this kind of case, where the protocol version causes very visible effects and requires application support, it may be safer to pin to a specific version - users can always explicitly opt into v2 if they want to (and have made the needed changes to support the new messages types).

Reply:
They wouldn't break immediately. They could get messages they don't expect for streaming transactions but the implementation could and probably should just log and ignore everything it doesn't know.
Beginners could also be surprised that they don't get the newest protocol features to work until we tell them that they need to explicitly turn on the newest protocol version.
Personally I would rather advise people to manually pin the protocol version and ignore messages they don't know, than always giving the most ancient protocol version to people who don't know it better but we can also do it the other way around and advise people to always set the protocol version they intend to support.
I am also a bit afraid that PostgreSQL will start dropping support for ancient protocol versions at some point but they tend to be pretty conservative with such things and it would probably give an exception immediately when starting replication which is the most benign problem you can have in that case.
As you like.

Reply:
There's even more to it.
For PG14 you can turn on different things:

1. Binary replication values with the `binary` option (for **both** protocol V1 and V2)
2. Logical decoding messages with the `messages` option (for **both** protocol V1 and V2)
3. Streaming of in-progress transactions with the `streaming` option (for protocol V2  **only**)

Only if you explicitly enable streaming (which is only possible for protocol V2) **all** messages will have the TransactionXid field. If you don't explicitly enable it, protocol V2 will behave like protocol V1.

Reply:
> They could get messages they don't expect for streaming transactions but the implementation could and probably should just log and ignore everything it doesn't know.

Sure, but that assumes the client application is well-written and properly ignores unknown messages. There's also the theoretical possibility that an existing message format would change across protocols, in which case there's no option to ignore.

>  Beginners could also be surprised that they don't get the newest protocol features to work until we tell them that they need to explicitly turn on the newest protocol version.

Isn't this the case with this PR already? As far as I can tell, unless I explicitly provide a protocol version to PgOutputReplicationOptions, it defaults to 1... If so, then the only change I'm proposing is to disallow providing null (float) as the protocol version.

Another way to look at this, is that PostgreSQL (apparently) requires clients to specify which protocol versions they want - they can't tell PostgreSQL "use the highest you can". If that is so, then it could be argued that Npgsql should expose exactly the same thing: require users to always provide the protocol version they want (that would even imply removing the default to 1 - which is probably not a bad idea).

(note that difference with the general, non-replication PG protocol - while using replication, protocol messages (and therefore the version) are directly exposed to users, the general PG protocol is an internal detail in the driver that isn't supposed to leak to the user, at least not explicitly).

Reply:
> If that is so, then it could be argued that Npgsql should expose exactly the same thing: require users to always provide the protocol version they want (that would even imply removing the default to 1 - which is probably not a bad idea).

Yes, that's probably the best thing to do. Let's not provide a default where PostgreSQL does not.
I'll do that.